### ISSUE TITLE
Proposal: Post-Render Hooks for Helm

### ISSUE THREAD

**Issue by thomastaylor312** [helm/helm#7260]
EDIT: I have kept all original text (with strikethrough) but have made modifications based on feedback

### tl;dr
If you don't want to read the whole thing, the basic gist is that a
post-renderer can be an arbitrary binary that must accept Kubernetes manifests
on STDIN and return 0 with valid Kubernetes manifests on STDOUT.

### Background

Many Helm users have been asking for a way to template charts in different ways
since the inception of Helm 2. Furthermore, many charts expose large numbers of
configuration options that only advanced users need and even then, there are
still tweaks that many users have to do that require forking a chart and
reducing reuse of the community maintained chart. At Helm Summit 2019, some of
the folks from Replicated gave a lightning talk on using Kustomize for their
last mile configuration of their charts. This led to the discussion and work for
this proposal.

### Problems

There are various problems I considered when writing this proposal. The biggest
of these are listed below:

- Using different templating engines has been discussed in the past, 
  but it would mean that charts would need to specify templating
  engine dependencies. This is less than ideal and will likely lead to less
  reuse and a poor user experience. Hence this solution
- ~How do we prevent accidental "shoot-self-in-foot" issues where you don't
  specify a renderer or the wrong renderer?~ No longer relevant here, but it was considered
- What is the best way to expose extra configuration without requiring it for a
  chart and while maintaining flexibility/extensibility?

### Solution
This solution focuses on keeping it simple, while providing room for all types
of additional configuration and validation. It also allows users to have this
extra functionality while maintaining their releases through the normal Helm CLI or SDK
(e.g. still being able to do upgrades and uninstalls)

#### Design

This option will be exposed through a CLI flag specified in the next section. For those using the Go SDK, we will expose the following interface that can be implemented however desired and passed to the install and upgrade clients:

```go
type PostRenderer interface {
	// Run expects a single buffer filled with Helm rendered manifests. It
	// expects the modified results to be returned on a separate buffer or an
	// error if there was an issue or failure while running the post render step
	Run(renderedManifests *bytes.Buffer) (modifiedManifests *bytes.Buffer, err error)
}
```

There will be an "exec" implementation provided with the new `postrender` package which will be used by the CLI to shell out to a post-rendering binary

##### Configuration
To expose this functionality in the `helm` CLI, we will add a new flag called
`--post-renderer` that will allow a user to specify the path to a binary to use
for their post-renderer. The renderer must be specified by the user every time they want to upgrade their release ~From a Go SDK and data storage standpoint, we will add
a new field called `Renderers` to the `Release` object. Currently this object
only looks like below:~

```Go
type Renderers struct {
    Post string
}
```

~This should allow us to do non-breaking feature adds if we decide to add more
renderers in the future, though that is currently unlikely.~

If the given string is an absolute path, it will use the binary located there.
Otherwise it will check first in your Helm plugins directory as it will make it
easy for people to distribute renders using the Helm plugins management. If it
is not found in your plugins, it will try to resolve the full path or check if
the binary exists in your `$PATH`. 

~NOTE: Post-rendering can be specified at install time, or can be added to an
existing release with an upgrade~

##### Rendering, Execution, and Failure
Helm will then render the charts using the standard Go template language and the
specified user values. Once this is done, but before parsing the manifests into
Kubernetes objects, `helm` will execute the binary and pass the charts on STDIN.
It will expect a 0 exit code with valid Kubernetes manifests on STDOUT. The
modified manifests will be validated against the Kubernetes API once they are
parsed. If the plugin has an error (or in the case of validation, fails
validation) it should return a non-zero exit code, in which case the release
will error out and fail.

~##### Data and "shoot-self-in-foot" protection~

~In order to prevent accidental overwrite or modification of releases, Helm will
provide a super basic check for all releases using post renders. If a user has
specified a post-renderer for a release, the name of that renderer is recorded
and they must use a renderer of the exact same name for the future life of the
release. If a different renderer is specified or no renderer is given during an
upgrade, the release will fail. This is meant to be basic protection only as
trying to handle varying versions or types of renderers is out of scope for the
Helm project.~

~NOTE: If desired, we could add a flag that allows a user to override a set
renderer. This would allow you to remove the need for a post-renderer from a
release or to switch the renderer you are using~

##### Security concerns

Obviously, the ability to inject data brings up a possible security concern. It
is not possible for Helm to guarantee the safety of every possible
post-renderer, so users will need to validate that their renderers are safe

#### Targeted use cases

There are 2 targeted use cases considered when designing this proposal:
- Anyone wanting to run compliance and/or security type scans: This is a very
  common requirement at large enterprises and allows organizations to inject
  common security functionality or perform scans as part of rendering
- Last mile configuration: As mentioned above, many different parts of a
  manifest like taints and tolerations are custom to each cluster. This proposal
  allows for those types of configurations to be added using tools such as
  Kustomize

#### Example
Attached to this proposal is a [tarball](https://github.com/helm/helm/files/3975506/kustomize-example.tar.gz) containing a simple Kustomize example for
use with the `stable/wordpress` chart (although it could be easily modified to
work with anything). To give this a try, you'll need to build the PR found
[here](https://github.com/helm/helm/pull/7259
) and expand the tarball into a directory. You will also need to make
sure you have the latest version of `kustomize` installed and available in your
`PATH`. Change directories into the directory you just expanded and run the
following commands:

```shell
$ /path/to/built/pr/helm install --post-renderer ./kustomize stable/wordpress --debug
```

You should be able to see in the output of the manifests that it set an
environment variable called `FOO` in the wordpress deployment. Please feel free
to take this and try it out with some more complex examples.

**Comment by joelanford** [comment]
@thomastaylor312 This could be useful for the issue I raised in #5826.

I'm curious what your thoughts would be on making the Helm Go SDK a bit more extensible, such that there is an exported `Renderer` interface. It would enable external users of the Go SDK to inject their own implementation of the renderer.

Then, the helm CLI could use a `PluginRenderer` and an `ExecutableRenderer` (or a combined one) that each implement the `Renderer` interface, and other users of the Helm Go SDK could implement and compile in their own.

As mentioned in #5826, my use case for this is to programmatically inject owner references into rendered manifests as part of an operator that reconciles releases in a cluster. Currently, in Helm 3 I'm not aware of a way to do this, and this was something that was possible in Helm 2.

**Comment by thomastaylor312** [comment]
That is a good point about the SDK. I'll take a shot at it after the holidays

**Comment by jlegrone** [comment]
> Different templating engines mean that charts would need to specify templating
engine dependencies. This is less than ideal and will likely lead to less
reuse and a poor user experience

This is an important point. Multiple users working on the same helm release would need to coordinate to ensure that they are using the same version of the renderer in order to get the same outputs. One way to avoid this would be to use a container-based plugin system, where a renderer is an OCI image specified in `Chart.yaml`. This would come at the cost of needing a docker daemon running, but would allow charts to remain portable and hermetic.

For example:

```yaml
apiVersion: v1
appVersion: "1.8.0"
description: My Example App
name: myapp
version: 2.1.3
renderers:
  - image: docker.io/helm/kustomize-render@sha256:somechecksum
  - image: example.com/helm-render/validate-image-signatures@sha256:somechecksum
```

Post-render images would handle manifests on some standard paths, eg. read `/helm/input-manifest.yaml` and write to `/helm/output-manifest.yaml`, and use stdout/stderr for logs and error messages which would be surfaced by the helm CLI.

**Comment by thomastaylor312** [comment]
@jlegrone So I haven’t updated the issue to take our discussion in the last dev call of 2019, but everyone was in favor of removing the checks that I have in code (they were a little too much). So I am all up for ways to make things more repeatable with it. However, I want SDK and other scripting users to not need something like a docker container. 

Another idea I just had (just spitballing here) is to add the name and SHA of the renderer to the `Chart.yaml` so it always checks against that SHA to make sure you are using the correct version. You can then update that SHA when you want to update the renderer. But like I said, that is a totally new idea we'd have to run by people

**Comment by jlegrone** [comment]
> Another idea I just had (just spitballing here) is to add the name and SHA of the renderer to the Chart.yaml so it always checks against that SHA

We would need to account for different platforms and multiple versions required by different charts, which might be a little tricky. But this would be valuable.

> I want SDK and other scripting users to not need something like a docker container.

I agree the downsides to requiring a docker daemon are huge. Also spitballing here, it would be interesting to combine something like https://github.com/wasmerio/go-ext-wasm and https://github.com/engineerd/wasm-to-oci to get the same benefits of portability while still only requiring the helm binary to be installed. Of course this comes with the downside of requiring end users to build + package their own webassembly binaries.

**Comment by thomastaylor312** [comment]
@joelanford I have updated the proposal and the associated PR to have a publicly exposed `PostRenderer` interface that should make it much easier to consume and use it with the SDK. Let me know what you think.

As for anyone else interested: I have updated the proposal with feedback I have received from members of the community. At this point, my one big concern has to do with @jlegrone's feedback. With this implementation style, we are putting the burden of repeatability on the end user to ensure they are using the correct post-renderer every time. Given the value add and advanced audience this feature is intended for, putting the burden on them to use the correct renderer could be ok. I need feedback either way and am looking for any possible options that don't involve large things like Docker or WASM (as much as WASM does interest me)

**Comment by jlegrone** [comment]
> Given the value add and advanced audience this feature is intended for, putting the burden on them to use the correct renderer could be ok.

My concern is that while this would be considered an advanced feature for a chart author, the onus is actually on the end user of the chart to have the correct rendering toolchain installed. This is generally fine for internal use cases, but could be a real issue for popular public charts.

**Comment by joelanford** [comment]
@thomastaylor312 The `PostRenderer` interface in the PR looks like it will be exactly what I need. Looking forward to this. Thanks!

**Comment by thomastaylor312** [comment]
@jlegrone I don't think it would be a problem for popular public charts. They shouldn't have to modify anything. This would be purely for the end user to decide if they want to run a post render. A public chart should never have a dependency on a specific post-renderer either. They could recommend one (like kustomize) for last mile config, but it shouldn't be a primary part of the chart. That is why I went this route instead. I didn't want to get involved with the mess of having renderer dependencies to install a chart. 

Hopefully that makes more sense. Let me know if it wasn't

**Comment by jlegrone** [comment]
> They could recommend one (like kustomize) for last mile config, but it shouldn't be a primary part of the chart.

Ah, I may have misunderstood the design. If invoking post-render is only possible via a Helm CLI flag (and not some config in `Chart.yaml`) then that addresses my concerns. 😄

**Comment by thomastaylor312** [comment]
As an FYI to all, we are going to keep this proposal open for 2 more weeks to give anyone from the community time to comment on it. Currently most of the maintainers are good with this proposal and those who have commented seem to be ok with things as well

**Comment by AndiDog** [comment]
- This should support empty input, e.g. using a completely different templating language tool instead of one that transforms YAML/manifests. For that purpose, you might want to consider allowing `templates/*.yaml` to be nonexistent (if that's not the case yet). Contrived example: `helm template --post-renderer my-handlebars-template-post-renderer` on a chart with only `templates/*.hbs` files. And as the example says: please not only support this in `helm install`, but also in `helm template` for those who don't use Helm for installation but only for manifest rendering.
- Consider setting helpful environment variables (I had Skaffold's [custom build script](https://skaffold.dev/docs/pipeline-stages/builders/custom/) support in mind here)
- Support multiple post renderers (repeated `--post-renderer`, called in order of CLI arguments). As the proposal says, it can be used for various purposes (compliance checks, manifest transformation, creation of manifests from non-YAML templates, etc.), so different tools might be desired.

**Comment by thomastaylor312** [comment]
@AndiDog Thank you for the feedback! One question concerning this point:

> Support multiple post renderers (repeated --post-renderer, called in order of CLI arguments). As the proposal says, it can be used for various purposes (compliance checks, manifest transformation, creation of manifests from non-YAML templates, etc.), so different tools might be desired.

So this was an explicit decision on my part. From the CLI perspective, having multiple post renders would add a lot of overhead of having to shell out to each thing, pull the buffer back in from stdout, then shelling out again for each one. My thinking was that if people wanted to use multiple post-renderers, they'd do it with a script (called something like `post-render-pipeline.sh`):

```shell
cat <&0 | thing one | thing two | thing three
```

Thoughts on that approach?

**Comment by AndiDog** [comment]
I agree, that would work as well. For my idea of using the script to generate manifests in the first place (input ignored), it could be something like:

```
#!/bin/sh
>&2 (
    tmpfile="$(mktemp ...)"
    my-fancy-template-tool ...my-cool-template-inputs... >"${tmpfile}"
    kubeval ... >&2
)
cat "${tmpfile}"
rm "${tmpfile}
```

**Comment by hickeyma** [comment]
@thomastaylor312 Proposal looks good to me. I think it is clear and easy to add a post renderer like `Kustomize`.   👍

**Comment by AndiDog** [comment]
This also solves the issue multiple people presented at Helm Summit 2019 – if someone wants to _maintain_ custom patches for upstream Helm charts, no more custom tooling is needed, but `--post-renderer` can be used with kustomize/yq/... to make any final changes to the manifest. It opens quite a few possibilities 😀

**Comment by thomastaylor312** [comment]
Ok, I have added support for `--post-renderer` on `helm template` as well. I am marking the PR as ready for review so we can accept and close this proposal

**Comment by thomastaylor312** [comment]
For those curious, the docs PR can be found here: https://github.com/helm/helm-www/pull/479

**Comment by kfox1111** [comment]
> > Different templating engines mean that charts would need to specify templating
> > engine dependencies. This is less than ideal and will likely lead to less
> > reuse and a poor user experience
> 
> This is an important point. Multiple users working on the same helm release would need to coordinate to ensure that they are using the same version of the renderer in order to get the same outputs. One way to avoid this would be to use a container-based plugin system, where a renderer is an OCI image specified in `Chart.yaml`. This would come at the cost of needing a docker daemon running, but would allow charts to remain portable and hermetic. This would also make it possible to specify arguments to renderers and pipeline multiple renderers without needing to pass tons of CLI flags to helm.
> 
> For example:
> 
> ```yaml
> apiVersion: v1
> appVersion: "1.8.0"
> description: My Example App
> name: myapp
> version: 2.1.3
> renderers:
> - image: docker.io/helm/kustomize-render@sha256:somechecksum
>   mounts:
>   - chartPath: kustomize.yaml # relative path to file within chart directory
>     mountPath: /helm/kustomize.yaml # path at which to mount file in the renderer image
>   - image: example.com/helm-render/validate-image-tags@sha256:somechecksum
>     env:
>     - name: BLACKLIST
>       value: 'latest,stable'
> ```
> 
> Post-render images would handle manifests on some standard paths, eg. read `/helm/input-manifest.yaml` and write to `/helm/output-manifest.yaml`, and use stdout/stderr for logs and error messages which would be surfaced by the helm CLI.

I like this idea. But why assume you need a docker daemon? Helm has a Kubernetes cluster backing it. You could spawn the container in k8s, pass in stdin, get stdout, and then delete it from k8s.

**Comment by thomastaylor312** [comment]
This has been merged and will be available in 3.1
